include/boost/math/distributions/kolmogorov_smirnov.hpp

```diff
@@ -0,0 +1,494 @@
+#include <boost/math/distributions/detail/common_error_handling.hpp>
+#include <boost/math/special_functions/jacobi_theta.hpp>
+#include <boost/math/tools/tuple.hpp>
+#include <boost/math/tools/roots.hpp> // Newton-Raphson
+#include <boost/math/tools/minima.hpp> // For the mode
+
+// #include <utility>
+
+namespace boost { namespace math {
+
+namespace detail {
+template <class RealType>
+inline RealType kolmogorov_smirnov_quantile_guess(RealType p) {
+    // Choose a starting point for the Newton-Raphson iteration
+    if (p > 0.9)
+        return RealType(1.8) - 5 * (1 - p);
+    if (p < 0.3)
+        return p + RealType(0.45);
+    return p + RealType(0.3);
+}
+
+// d/dk (theta2(0, 1/(2*k*k/M_PI))/sqrt(2*k*k*M_PI))
+template <class RealType, class Policy>
+RealType kolmogorov_smirnov_pdf_small_x(RealType x, RealType n, const Policy&) {
+    BOOST_MATH_STD_USING
+    RealType value = RealType(0), delta = RealType(0), last_delta = RealType(0);
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    int i = 0;
+    RealType pi2 = constants::pi_sqr<RealType>();
+    RealType x2n = x*x*n;
+    if (x2n*x2n == 0.0) {
+        return static_cast<RealType>(0);
+    }
+    while (1) {
+        delta = exp(-RealType(i+0.5)*RealType(i+0.5)*pi2/(2*x2n)) * (RealType(i+0.5)*RealType(i+0.5)*pi2 - x2n);
+
+        if (delta == 0.0)
+            break;
+
+        if (last_delta != 0.0 && fabs(delta/last_delta) < eps)
+            break;
+
+        value += delta + delta;
+        last_delta = delta;
+        i++;
+    }
+
+    return value * sqrt(n) * constants::root_half_pi<RealType>() / (x2n*x2n);
+}
+
+// d/dx (theta4(0, 2*x*x*n/M_PI))
+template <class RealType, class Policy>
+inline RealType kolmogorov_smirnov_pdf_large_x(RealType x, RealType n, const Policy&) {
+    BOOST_MATH_STD_USING
+    RealType value = RealType(0), delta = RealType(0), last_delta = RealType(0);
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    int i = 1;
+    while (1) {
+        delta = 8*x*i*i*exp(-2*i*i*x*x*n);
+
+        if (delta == 0.0)
+            break;
+
+        if (last_delta != 0.0 && fabs(delta / last_delta) < eps)
+            break;
+
+        if (i%2 == 0)
+            delta = -delta;
+
+        value += delta;
+        last_delta = delta;
+        i++;
+    }
+
+    return value * n;
+}
+
+}; // detail
+
+template <class RealType = double, class Policy = policies::policy<> >
+    class kolmogorov_smirnov_distribution
+{
+    public:
+        typedef RealType value_type;
+        typedef Policy policy_type;
+
+        // Constructor
+    kolmogorov_smirnov_distribution( RealType n ) : n_obs(n)
+    {
+        RealType result;
+        detail::check_df(
+                "boost::math::kolmogorov_smirnov_distribution<%1%>::kolmogorov_smirnov_distribution", n_obs, &result, Policy());
+    }
+
+    RealType number_of_observations()const
+    {
+        return n_obs;
+    }
+
+    private:
+
+    RealType n_obs; // positive integer
```

Comment:
We generally use either a suffixed underscore or `m_` prefix for private member variables; I'm do `n_obs_` here.